streams/src/test/java/org/apache/kafka/streams/integration/OptimizedKTableIntegrationTest.java

```diff
@@ -0,0 +1,293 @@
+
+        produceValueRange(key, 0, 10);
+
+        assertThat("all messages in the first batch were processed in a timely manner",
+            semaphore.tryAcquire(numMessages, 60, TimeUnit.SECONDS));
+
+        assertThat("no restore has occurred", restoreStartOffset.get(), is(equalTo(-1L)));
+    }
+
+    @Test
+    public void shouldApplyUpdatesToStandbyStore() throws InterruptedException, ExecutionException {
+        final int batch1NumMessages = 100;
+        final int batch2NumMessages = 100;
+        final int key = 1;
+        final Semaphore semaphore = new Semaphore(0);
+
+        final StreamsBuilder builder = new StreamsBuilder();
+        builder
+            .table(INPUT_TOPIC_NAME, Consumed.with(Serdes.Integer(), Serdes.Integer()),
+                Materialized.<Integer, Integer, KeyValueStore<Bytes, byte[]>>as(TABLE_NAME)
+                    .withCachingDisabled())
+            .toStream()
+            .peek((k, v) -> semaphore.release());
+
+        final KafkaStreams client1 = createClient(builder, streamsConfiguration());
+        final KafkaStreams client2 = createClient(builder, streamsConfiguration());
+        final List<KafkaStreams> clients = Arrays.asList(client1, client2);
+
+        final AtomicLong restoreStartOffset = new AtomicLong(-1L);
+        clients.forEach(client -> {
+            client.setGlobalStateRestoreListener(createTrackingRestoreListener(restoreStartOffset));
+            client.start();
+        });
+        waitForClientsToEnterRunningState(clients, 60, TimeUnit.SECONDS);
+
+        produceValueRange(key, 0, batch1NumMessages);
+
+        assertThat("all messages in the first batch were processed in a timely manner",
+            semaphore.tryAcquire(batch1NumMessages, 60, TimeUnit.SECONDS));
+
+        final ReadOnlyKeyValueStore<Integer, Integer> store1 = client1
+            .store(TABLE_NAME, QueryableStoreTypes.keyValueStore());
+
+        final ReadOnlyKeyValueStore<Integer, Integer> store2 = client2
+            .store(TABLE_NAME, QueryableStoreTypes.keyValueStore());
+
+        final boolean client1WasFirstActive;
+        if (store1.get(key) != null) {
+            client1WasFirstActive = true;
+        } else {
+            assertThat("data from the job was sent to the store", store2.get(key) != null);
+            client1WasFirstActive = false;
+        }
+
+        assertThat("no restore has occurred", restoreStartOffset.get(), is(equalTo(-1L)));
+        assertThat("current value in store should reflect all messages being processed",
+            client1WasFirstActive ? store1.get(key) : store2.get(key), is(equalTo(batch1NumMessages - 1)));
+
+        if (client1WasFirstActive) {
+            client1.close();
+        } else {
+            client2.close();
+        }
+
+        final int totalNumMessages = batch1NumMessages + batch2NumMessages;
+
+        produceValueRange(key, batch1NumMessages, totalNumMessages);
```

Comment:
Before we produce new messages, should we verify that the store was migrated correctly and contains `store.get(key), is(equalTo(batch1NumMessages - 1))` ?

Reply:
Yes, we can do that with a time limited loop that queries for that state.

Reply:
@mjsax You may have uncovered yet another bug (BUG 6?) in optimized graph failover! I cannot seem to get any value for the key after failover. So, looks like this may involve yet more significant investigation. Thanks for the idea!

Reply:
This test fails with non-optimized graphs too. I hope I'm doing something wrong...

Reply:
*whew*... Apparently 10 seconds was not long enough to wait for the store to become available after failure. If I wait a bit longer it finally becomes available. Seems long though? This is the case for both non-optimized and optimized...

Reply:
If case of failure, we detect the failure only after `session.timeout.ms` (default 10 seconds) hit -- to speed up the test, we could decrease the session timeout via `StreamsConfig`

Reply:
I had the same exact thought. My only reservation was tweaking config too much might start to deviate too much from expected use. In other words, I'm totally for this if have a reasonable value in mind. I was thinking about 1s?

Reply:
We can try that -- but we need to reduce `HEARTBEAT_INTERVAL_MS_CONFIG`, too, for this case -- its default is 3 seconds and it must be smaller than session timeout.

Checking some existing tests, we have some that set them to 500ms / 1sec. Some other test reduce session timeout to 2 or 5 seconds...